common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/Indexer.scala

```diff
@@ -25,14 +24,53 @@ trait Indexable[T] {
   def version(document: T): Int
 }
 
+/**
+  * Versioning:
+  *
+  * When the merger makes the decision to merge some works, it modifies
+  * the content of the affected works. Despite the content of these works
+  * being modified, their version remains the same. However, the merger
+  * attaches the number of merged sources (`nMergedSources`) to the target
+  * work and may choose to redirect some of the source works.
+  *
+  * When we ingest those works, we need to make sure that the merger
+  * modified works are never overridden by unmerged works for the same
+  * version still running through the pipeline.
+  *
+  * We also need to make sure that, if a work is modified by a source in
+  * such a way that it shouldn't be merged (or redirected) anymore, the
+  * new unmerged version is ingested and never replaced by the previous
+  * merged version still running through the pipeline.
+  *
+  * We can do that by ingesting works into Elasticsearch with a document
+  * version derived from a combination of work version, number of merged
+  * sources and work type.
+  *
+  * More specifically, by multiplying the work version by 10, we make sure
+  * that a new version of a work always wins over previous versions
+  * (merged or unmerged, provided that there are fewer than 10 merged
+  * sources, which we think is a reasonable assumption).
+  *
+  * We make sure that a merger modified work always wins over other works
+  * with the same version, by adding the number of merged sources to
+  * work.version * 10.
+  *
+  * The exact same logic is applied to image versions, using both the
+  * merge state and the version numbers of the image's SourceWorks.
+  */
```

Comment:
If I've read this correctly, I think there's still a scenario where the correct version wouldn't make it to the index:
-work A version 1 is linked to work B version 1
- work A version 1 is unlinked from work B
-work A version1 is linked to work C <-- this would not make it to the index 
In practice this may be a convoluted example that never happens though..

Reply:
I think you're probably right (and I think there are probably some similar but different problem scenarios too). I'm kind of willing to let these go for now unless we see them happening as they seem to be bugs that've been there all along, what do you think?

Reply:
Actually, in the above scenario, how would unlinked workA make it to the index? the number of merged sources goes down in that case

Reply:
I think the above scenario is probably convoluted and I wouldn't worry. I would put a comment for whoever might need to investigate this in the future though. 
But I think the ability to unlink is a requirement, and I'm not entirely sure I understand how it works

Reply:
You're right, and I feel similarly confused by it 🙃  - we explicitly say (and have always said) that a non-redirected work _doesn't_ override a redirected one, but what if it's become non-redirected because it's been unlinked? Perhaps we are relying on a change to the source data (and hence the version in the source) to cause that, and because changes to merger rules usually (always) require a reindex it doesn't matter. I'll add a comment!

Reply:
Yeah, I'm quite confused because now I can't understand how this ever worked. I don't think we can always assume that if something gets unlinked it's a new version in the sourcedata. For example, work A might be unlinked from work B because of a change in workB sourcedata. So workB advances in version, but workA doesn't and because the number of linked works is lower, the unlinked one never makes it in the index

Reply:
so I think how this used to sort of work is that previously we used the `merged` flag set by the merger to determine the version. Which would be set to `true` even when unlinking because that gets set always if the merger finds a target and potential redirects regardless of whether any of the rules got applied. So the final es version used to be the same for the linked and unlinked version of workA which would mean the second to arrive would make it to the index because the es versioning is greater _or equal_.

Bottom line is, I don't think this case works now. and I _think_ this may be a problem but admittedly the way it used to work is way too obscure

Reply:
Ohh I see the difference now, good deduction. I will try keeping it like this (since it's more explicit / less obscure) and see what happens I think